pros/serial/devices/vex/v5_device.py

```diff
@@ -514,14 +531,22 @@ def read_file(self, file: typing.IO[bytes], remote_file: str, vid: int_str = 'us
                     f'consider downloading directly with a wire.', abort=True, default=False)
 
         max_packet_size = ft_meta['max_packet_size']
-        with ui.progressbar(length=file_len, label='Downloading {}'.format(remote_file)) as progress:
+        if silent:
             for i in range(0, file_len, max_packet_size):
                 packet_size = max_packet_size
                 if i + max_packet_size > file_len:
                     packet_size = file_len - i
                 file.write(self.ft_read(addr + i, packet_size))
-                progress.update(packet_size)
                 logger(__name__).debug('Completed {} of {} bytes'.format(i + packet_size, file_len))
+        else:
+            with ui.progressbar(length=file_len, label='Downloading {}'.format(remote_file)) as progress:
+                for i in range(0, file_len, max_packet_size):
+                    packet_size = max_packet_size
+                    if i + max_packet_size > file_len:
+                        packet_size = file_len - i
+                    file.write(self.ft_read(addr + i, packet_size))
+                    progress.update(packet_size)
+                    logger(__name__).debug('Completed {} of {} bytes'.format(i + packet_size, file_len))
```

Comment:
Is there a better way to get silent operation without needing to duplicate this logic? Perhaps we can propagate silent into pros.common.ui.progressbar or something similar.

Reply:
I don't think making the progressbar silent would be the way to go since that is the entire point of the progress bar. Maybe just put the code inside the for loop (minus the progress.update part) in a separate function and call it every time in with the for loop? That way there's no major change to the progress bar code and gets rid of code reuse and then progessbar.update can be called in the one that uses the progressbar.

Reply:
something like this would probably work. does the progressbar show up if you never update it? I can't remember
```py
# e.g.: self._write_file_packets(file, max_packet_size, progress.update)
def _write_file_packets(file, max_packet_size, progress_cb=None):
    file_len = len(file)
    for i in range(0, file_len, max_packet_size):
        packet_size = max_packet_size
        if i + max_packet_size > file_len:
            packet_size = file_len - i
        file.write(self.ft_read(addr + i, packet_size))
        if progress_cb is not None:
            progress_cb(packet_size)
        logger(__name__).debug(f'Completed {i + packet_size} of {file_len} bytes')
```

Reply:
The progressbar still shows up if you never update it - just stays at 0% -, but doesn't show if you just keep the ```with ui.progressbar(...)``` out of it. The function you made works with a few modifications: ```file_len``` has to be passed in since it's grabbed from the file metadata and ```addr``` needs to be a parameter as well.
This worked for me:

```python
def _write_file_packets(self, file, file_len, max_packet_size, addr, progress_cb=None):
        for i in range(0, file_len, max_packet_size):
            packet_size = max_packet_size
            if i + max_packet_size > file_len:
                packet_size = file_len - i
            file.write(self.ft_read(addr + i, packet_size))
            if progress_cb is not None:
                progress_cb(packet_size)
            logger(__name__).debug(f'Completed {i + packet_size} of {file_len} bytes')
```

And inside of ```read_file```:
```python
if silent:
            self._write_file_packets(file, file_len, max_packet_size, addr)
        else:
            with ui.progressbar(length=file_len, label='Downloading {}'.format(remote_file)) as progress:
                self._write_file_packets(file, file_len, max_packet_size, addr, progress.update)
```

Reply:
Right, okay then. Was just toying around with the idea of duplicating as little as possible. One small nit: prefer f-string literals wherever possible